crates/nu-cli/src/completions/operator_completions.rs

```diff
@@ -2,155 +2,255 @@ use crate::completions::{
+    vec![
+        operator_to_item(ast::Assignment::Assign),
+        operator_to_item(ast::Assignment::ConcatenateAssign),
+    ]
+}
+
+fn valid_int_ops() -> Vec<OperatorItem> {
+    let mut ops = valid_float_ops();
+    ops.extend(bit_ops());
+    ops
+}
+
+fn valid_float_ops() -> Vec<OperatorItem> {
+    let mut ops = valid_value_with_unit_ops();
+    ops.push(operator_to_item(ast::Math::Pow));
+    ops
+}
+
+fn valid_string_ops() -> Vec<OperatorItem> {
+    let mut ops: Vec<OperatorItem> = Comparison::iter().map(operator_to_item).collect();
+    ops.push(operator_to_item(ast::Math::Concatenate));
+    ops.push(OperatorItem {
+        symbols: "like".into(),
+        description: Comparison::RegexMatch
+            .get_message()
+            .unwrap_or_default()
+            .into(),
+    });
+    ops.push(OperatorItem {
+        symbols: "not-like".into(),
+        description: Comparison::NotRegexMatch
+            .get_message()
+            .unwrap_or_default()
+            .into(),
+    });
+    ops
+}
+
+fn valid_list_ops() -> Vec<OperatorItem> {
+    let mut ops = collection_comparison_ops();
+    ops.push(operator_to_item(ast::Math::Concatenate));
+    ops
+}
+
+fn valid_binary_ops() -> Vec<OperatorItem> {
+    let mut ops = number_comparison_ops();
+    ops.extend(bit_ops());
+    ops.push(operator_to_item(ast::Math::Concatenate));
+    ops
+}
+
+fn valid_bool_ops() -> Vec<OperatorItem> {
+    let mut ops: Vec<OperatorItem> = ast::Boolean::iter().map(operator_to_item).collect();
+    ops.extend(common_comparison_ops());
+    ops
+}
+
+fn valid_value_with_unit_ops() -> Vec<OperatorItem> {
+    let mut ops = number_comparison_ops();
+    ops.extend(math_ops());
+    ops
+}
+
+fn ops_by_value(value: &Value, mutable: bool) -> Vec<OperatorItem> {
+    let mut ops = match value {
+        Value::Int { .. } => valid_int_ops(),
+        Value::Float { .. } => valid_float_ops(),
+        Value::String { .. } => valid_string_ops(),
+        Value::Binary { .. } => valid_binary_ops(),
+        Value::Bool { .. } => valid_bool_ops(),
+        Value::Date { .. } => number_comparison_ops(),
+        Value::Filesize { .. } | Value::Duration { .. } => valid_value_with_unit_ops(),
+        Value::Range { .. } | Value::Record { .. } => collection_comparison_ops(),
+        Value::List { .. } => valid_list_ops(),
+        _ => common_comparison_ops(),
+    };
+    if mutable {
+        ops.extend(match value {
+            Value::Int { .. }
+            | Value::Float { .. }
+            | Value::Filesize { .. }
+            | Value::Duration { .. } => numeric_assignment_ops(),
+            Value::String { .. } | Value::Binary { .. } | Value::List { .. } => {
+                concat_assignment_ops()
+            }
+            _ => vec![operator_to_item(ast::Assignment::Assign)],
+        })
+    }
+    ops
+}
+
+fn is_expression_mutable(expr: &Expr, working_set: &StateWorkingSet) -> bool {
+    let Expr::FullCellPath(path) = expr else {
```

Comment:
This doesn't account for `$env.FOO`

Reply:
True, what do you suggest, maybe special checks for `env`? `$nu` is immutable, right?

Reply:
I see special treatment for `ENV_VARIABLE_ID` somewhere in parser too, makes me wonder why not mark it as mutable in the first place.

Can't find the right place to make that happen.